Edit cluster template docs

catherineluse opened the pull request:
This PR contains feature documentation about cluster templates, corresponding to this issue: https://github.com/rancher/docs/issues/1478

This PR also continues the work done in https://github.com/rancher/docs/pull/1622.

In this set of changes, I have done a few things:

- I added some information about cluster templates to the documentation about Rancher-launched Kubernetes clusters. This is so that the template docs are more discoverable for someone who wants to create a new cluster.
- I added some clarification about setting the Kubernetes version ending with .x in a cluster template.
- Reorganized sections and renamed headers for clarity and discoverability
- Tried to be super clear (even to the point of redundancy) about the RBAC and restrictions on clusters created with templates

deniseschannon commented on content/rancher/v2.x/en/user-settings/cluster-templates/cluster-templates-and-hardware/_index.md:
```diff
@@ -0,0 +1,70 @@
+
+Terraform is a server provisioning tool. It uses infrastructure-as-code that lets you create almost every aspect of your infrastructure with Terraform configuration files. It can automate the process of server provisioning in a way that is self-documenting and easy to track in version control.
+
+This section focuses on [Terraform with the Rancher 2 provider](https://www.terraform.io/docs/providers/rancher2/), which is a recommended option to standardize the hardware for your Kubernetes clusters. It allows you to quickly create a comprehensive, production-ready cluster.
```
I thought this terraform is what is used to bring up Rancher 2, but not the clusters within Rancher. Can you confirm @cloudnautique ?

catherineluse replied:
In Jason van Brackel's Ranchcast episode, he said you can use the Terraform provider to provision hardware, but you would need Terraform modules to create the entire cluster within Rancher. I should probably change the last sentence so it doesn't sound like the provider can do everything.

deniseschannon commented on content/rancher/v2.x/en/user-settings/cluster-templates/enforcement-and-overrides/_index.md:
```diff
@@ -0,0 +1,45 @@
+---
+title: Template Enforcement and Overrides
```
I feel as if enforcement and overrides should be in 2 different sections. They are completely different concepts

catherineluse replied:
split into two sections

deniseschannon commented on content/rancher/v2.x/en/user-settings/cluster-templates/owner-permissions/_index.md:
```diff
@@ -0,0 +1,72 @@
+
+If you create a template, you automatically become an owner of that template. 
+
+The owner of a template can revise it, share it, and delete it. The owner can also share the template with other users who have the **Create Cluster Templates** global role.
```
This should expand to cover what an owner can do for the template as well as the template revisions.

catherineluse replied:
revised the intro and added a link to the section about how template revisions work

deniseschannon commented on content/rancher/v2.x/en/user-settings/cluster-templates/owner-permissions/_index.md:
```diff
@@ -0,0 +1,72 @@
+- By editing the permissions of an [individual user](#allowing-a-user-to-create-and-revise-templates)
+- By changing the [default permissions of new users](#allowing-new-users-to-create-and-revise-templates-by-default)
+
+### Allowing a User to Create and Revise Templates
```
This permissions will change when https://github.com/rancher/rancher/issues/22645 is implemented.

catherineluse replied:
Changed passages about permissions

deniseschannon commented on content/rancher/v2.x/en/user-settings/cluster-templates/owner-permissions/_index.md:
```diff
@@ -0,0 +1,72 @@
+1. Click **Save.**
+
+**Result:** The user cannot create cluster templates. If this user is given owner privileges for a template, they won't be able to use those privileges.
+
```
Can we remove the extra line breaks?

catherineluse replied:
removed

catherineluse commented on content/rancher/v2.x/en/cluster-provisioning/rke-clusters/_index.md:
```diff
@@ -9,24 +9,44 @@ If you don't want to use a hosted Kubernetes provider, you can have Rancher laun
+Using Rancher, you can create pools of nodes based on a [node template]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/rke-clusters/node-pools/#node-templates). This node template defines the parameters you want to use to launch nodes in your cloud providers. The available cloud providers to create a node template are decided based on active [node drivers]({{< baseurl >}}/rancher/v2.x/en/cl ...
 
-- [Custom Nodes]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/rke-clusters/custom-nodes/):
+By default, if you have a pool of nodes, but one loses connectivity and goes inactive, rancher does not automatically spin up a new node until the inactive node is deleted from Rancher. As of Rancher v2.3, you can configure a node pool so that a node is deleted and replaced if it becomes inactive for a specified number of minutes.
```
Changed the paragraph to "As of Rancher v2.3, node pools have a node auto-replace feature that automatically deletes and replaces nodes that become inactive for a specified number of minutes."

catherineluse commented on content/rancher/v2.x/en/cluster-provisioning/rke-clusters/node-pools/_index.md:
```diff
@@ -5,19 +5,59 @@ aliases:
+A node template is the saved configuration for the parameters to use when provisioning nodes in a specific cloud provider. Rancher provides a nice UI to be able to launch these nodes and uses [Docker Machine](https://docs.docker.com/machine/) to provision these nodes. The available cloud providers to create node templates are based on the active node drivers in Rancher.
 
-Each node pool is assigned with a [node component]({{< baseurl >}}/rancher/v2.x/en/cluster-provisioning/#node-components) to specify how these nodes should be configured for the Kubernetes cluster.
+After you create a node template in Rancher, it's saved so that you can use this template again to create [node pools.](#node-pools) Node templates are bound to your login. After you add a template, you can remove them from your user profile.
```
removed link

catherineluse commented:
I have updated this PR with the following changes:

- Responded to Denise's feedback from September 4
- Moved docs from user settings to global settings
- Updated verbiage from "Cluster template" to "RKE template"
- Updated how permissions are handled (you need a global role to create a template, but you no longer need the global role to be an owner of an existing template)

This update addresses this issue about the UI changes https://github.com/rancher/docs/issues/1709

This PR should not be merged until the following two things are confirmed (@prachidamle  is finding out):
- Is the global role name going to change from "Create Cluster Templates" to "Create RKE Templates"?
- In the global settings about template enforcement, is the setting name going to change from `cluster-template-enforcement` to `rke-template-enforcement`?

This PR assumes those changes will be made. If not, I'll change them back.

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+
+Administrators can turn on template enforcement to require users to always use RKE templates when creating a cluster. This allows administrators to guarantee that Rancher always provisions clusters with specific settings.
+
+RKE template updates are handled through a revision system. If you want to change or update a template, you can create a new revision of the template. Then a cluster that was created with the older version of the template can be upgraded to the new template revision.
```
Small nit, ` If you want to change or update a template, you can create a new revision of the template. ` -> ` If you want to change or update a template, you create a new revision of the template. ` 

Since most of our docs are instructing someone what to do, we should avoid using the word `can` so it's more directive.

catherineluse replied:
removed "can"

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+
+RKE template updates are handled through a revision system. If you want to change or update a template, you can create a new revision of the template. Then a cluster that was created with the older version of the template can be upgraded to the new template revision.
+
+There are two types of settings in an RKE template: settings that can be overridden by the end user and settings that cannot. The difference is indicated by the **Allow User Override** toggle over each setting in the Rancher UI when the template is created.
```
I wouldn't call this 2 types of settings. It's more as if a setting can either be restricted to what the cluster owner chooses OR open for the end user to select the value.

catherineluse replied:
changed to 

In an RKE template, settings can be restricted to what the template owner chooses, or they can be open for the end user to select the value. The difference is indicated by the **Allow User Override** toggle over each setting in the Rancher UI when the template is created.

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+
+There are two types of settings in an RKE template: settings that can be overridden by the end user and settings that cannot. The difference is indicated by the **Allow User Override** toggle over each setting in the Rancher UI when the template is created.
+
+For the settings that cannot be overridden, the end user will not be able to directly edit them. The only way to edit these options would be to upgrade to a new revision of the same RKE template.
```
`The only way to edit these options would be to upgrade to a new revision of the same RKE template.` . -> `In order for a user to get different options of these settings, a RKE template owner would need to create a new revision of the RKE template, which would allow the user to upgrade and change that option.`

catherineluse replied:
updated

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+
+# YAML Customization
+
+If you define an RKE template as a YAML file, you can modify this [example RKE template YAML]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rke-templates/example-yaml).
```
We should add a note that the YAML in the RKE template uses the same customization when creating a RKE cluster, but since the yaml is within the context of a Rancher providisioned RKE cluster, the customization from the RKE docs needs to be nested under `rancher_kubernetes_engine` (you might need to confirm what this directive is) 

catherineluse replied:
explained in YAML customization section and example yaml page

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+- Install plugins on nodes that are deployed with a Kubernetes daemonset
+- Automatically set up namespaces, service accounts, or role binding
+
+To set add-ons, when creating the template, click **Edit as YAML.** Then use the `addons` or `addons_include` directive to set which YAML files are used for the add-ons.
```
If we're going to write about this, we should make sure it's also noted that it's nested within the `rancher_kubernetes_engine` directive.

catherineluse replied:
added paragraph

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/applying-templates/_index.md:
```diff
@@ -0,0 +1,31 @@
+1. Provide the cluster name and node template details as usual.
+1. To use an RKE template, under the **Cluster Options**, check the box for **Use an existing RKE template and revision.**
+1. Choose an existing template and revision from the dropdown menu.
+1. Optional: You can edit any settings that the RKE template owner marked as **Allow User Override** when the template was created.
```
Do you think we should note that if there are settings that you want to change, but don't have the option to, they'd need to contact a template owner to get a a new revision?

catherineluse replied:
Added sentence

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/applying-templates/_index.md:
```diff
@@ -0,0 +1,31 @@
+
+### Updating a Cluster Created with an RKE Template
+
+When the cluster owner creates a template, each setting has a switch in the Rancher UI that indicates if users can override the setting.
```
cluster owner -> template owner

catherineluse replied:
updated

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/applying-templates/_index.md:
```diff
@@ -0,0 +1,31 @@
+
+If a cluster was created from an RKE template, you can edit the cluster to update the cluster to a new revision of the template.
+
+You can't change the cluster to use a different RKE template. You can only update the cluster to a new revision of the same template.
```
We should note this so it sticks out more. and maybe put it in the beginning of using RKE templates.

catherineluse replied:
added this note here and in the beginning of the main RKE template doc

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/creating-and-revising/_index.md:
```diff
@@ -0,0 +1,104 @@
+
+**Result:** The template is updated.
+
+### Deleting a Template
```
can you delete revisions or only disable them?

catherineluse replied:
You can delete all revisions except the default revision. But even the default revision can be disabled. Adding it to the docs

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/creating-and-revising/_index.md:
```diff
@@ -0,0 +1,104 @@
+---
+title: Creating and Revising Templates
```
In this whole section, can you re-review how we reference how to use a template/revision? most of it references being able to be used for a new cluster, but any revision is actually for a new cluster or existing cluster that used that cluster template.

catherineluse replied:
added paragraph

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/creating-and-revising/_index.md:
```diff
@@ -0,0 +1,104 @@
+
+When you update an RKE template, you are creating a revision of the existing template. Clusters that were created with an older version of the template can be updated to match the new revision.
+
+You can't edit individual revisions. However, if you want to prevent a revision from being used to create a new cluster, you can [disable it.](#disabling-a-template-revision)
```
`Since you can't edit individual revisions of a template, in order to prevent a revision from being used, you can disable it. `

catherineluse replied:
updated

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/enforcement/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+This section also describes how template owners can define which settings in a template can be edited by end users.
+
+By default, only an administrator has the ability to create clusters without a template. All other users must use an RKE template when creating a cluster. Users can only create new templates if the administrator [gives them permission.]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rke-templates/creator-permissions/#allowing-a-user-to-create-templates)
```
I'd say that by default, any user in rancher can create clusters, but you can change the default so taht only admins have the ability.

catherineluse replied:
changed to

By default, any standard user in Rancher can create clusters, but you can change the default so that only an administrator has the ability to create clusters without a template. When you turn on temmplate enforcement, all standard users must use an RKE template when creating a cluster.

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/enforcement/_index.md:
```diff
@@ -0,0 +1,34 @@
+
+# Requiring New Clusters to Use an RKE Template
+
+You might want to require new clusters to use a template to ensure that any cluster launched by a non-administrator will use the Kubernetes and/or Rancher settings that are vetted by administrators.
```
instad of "non-administrative", use "standard user" and link to the standard user explanation (https://rancher.com/docs/rancher/v2.x/en/admin-settings/rbac/global-permissions/)

catherineluse replied:
updated

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/example-scenarios/_index.md:
```diff
@@ -0,0 +1,66 @@
+1. First, an administrator creates a template which specifies the Kubernetes version as 1.14 and marks all other settings as **Allow User Override**. 
+1. Then the administrator makes the template public and turns on template enforcement.
+
+**Result:** All Rancher users in the organization have access to the template. All users except for administrators are required to use a template when creating new clusters.
```
Maybe it should also say "All clusters created would be unable to pick a different Kubernetes version and therefore all clusters would be Kubernetes 1.14"

catherineluse replied:
changed result to:

**Result:** All Rancher users in the organization have access to the template. All new clusters created by [standard users]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rbac/global-permissions/) use Kubernetes 1.14 and they are unable to use a different Kubernetes version.

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/example-scenarios/_index.md:
```diff
@@ -0,0 +1,66 @@
+
+Let's say an organization has both basic and advanced users. Adminstrators want the basic users to be required to use a template, while the advanced users and administrators can create their clusters however they want.
+
+1. First, an administrator turns on [RKE template enforcement.]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rke-templates/enforcement/#requiring-new-clusters-to-use-a-cluster-template) This means that every non-administrator in Rancher will need to use an RKE template when they create a cluster.
```
non-administrator -> standard user

catherineluse replied:
updated

catherineluse commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -0,0 +1,117 @@
+
+# Configurable Settings
+
+RKE templates can be created in the Rancher UI or defined in YAML format. They can define all the same parameters that can be specified in Rancher’s cluster creation feature, including the following:
```
changed to "They can define all the same parameters that can be specified when you use Rancher to provision custom nodes or nodes from an infrastructure provider:"

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -98,24 +98,22 @@ RKE templates are designed to standardize Kubernetes and Rancher settings. If yo
 # YAML Customization
 
-If you define an RKE template as a YAML file, you can modify this [example RKE template YAML]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rke-templates/example-yaml).
+If you define an RKE template as a YAML file, you can modify this [example RKE template YAML]({{<baseurl>}}/rancher/v2.x/en/admin-settings/rke-templates/example-yaml). The YAML in the RKE template uses the same customization that Rancher uses when creating an RKE cluster, but since the YAML is located within the context of a Rancher provisioned cluster, you will need to nest the RKE template cust ...
```
The directive is actually `rancher_kubernetes_engine_config`. I didn't have great internet access when I had first wrote the comment, so I couldn't look it up.

deniseschannon commented on content/rancher/v2.x/en/admin-settings/rke-templates/_index.md:
```diff
@@ -98,24 +98,22 @@ RKE templates are designed to standardize Kubernetes and Rancher settings. If yo
-The RKE template configuration must be nested within the `rancher_kubernetes_engine` directive.
-
-For information on custom add-ons, refer to the [user-defined add-ons documentation.]({{<baseurl>}}/rke/latest/en/config-options/add-ons/user-defined-add-ons/)
+The RKE template configuration must be nested within the `rancher_kubernetes_engine` directive. To set add-ons, when creating the template, you will click **Edit as YAML.** Then use the `addons` directive to add a manifest, or the `addons_include` directive to set which YAML files are used for the add-ons. For more information on custom add-ons, refer to the [user-defined add-ons documentation.]( ...
```
The directive is actually rancher_kubernetes_engine_config.
